### PR TITLE
Add masonry layout grid

### LINKED ISSUE #378: 🐛[Bug] All PostCards have the same height - dynamic height required
### All PostCards have the same height - dynamic height required

max height must be defined

image

image

### Steps to reproduce the behavior
1. Login
2. See Cards
3. All same Height
4. ...
5. Profit

### Expected behavior

image

### Version & Environment
 Type: [Desktop]      
 - OS: [Windows]      
 - Browser: [Chrome Beta] 
 - Version https://github.com/Human-Connection/Human-Connection/commit/e4cfa2403a457032c10b449abeaf3082e1863111  

### Additional Context

Keyword: masonry grid

### PR DESCRIPTION
**Pull request by alina-beck**
## 🍰 Pullrequest
Use CSS grid to create a masonry layout – allowing posts to have different heights.

So instead of this:
Screenshot 2019-08-13 at 16 42 47

We should now see this:
Screenshot 2019-08-13 at 16 49 48

### Issues
- fixes #378
- relates #372
- waiting for PR: https://github.com/Human-Connection/Nitro-Styleguide/pull/131

### ToDo
- [x] update styleguide dependency
- [x] use styleguide components in project

### CODE REVIEW

**Comment by roschaefer** [comment]
@alina-beck I rebased your PR - you will need to `git fetch && git reset origin/add-masonry-grid --hard`

**Comment by alina-beck** [comment]
@roschaefer this is still a work in progress – I need the [styleguide PR](https://github.com/Human-Connection/Nitro-Styleguide/pull/131) merged and released, then I can finish this one. I still have a few plans. :wink: But thank you for your suggestions! The diagram is really helpful. 😍 

May I ask why you are force-pushing and rebasing instead of merging in master? I've only ever done it locally and thought it wouldn't be a good idea to do it on the remote repo... Now I'm curious to hear your reasons because I've seen you do it quite often!

File: webapp/components/PostCard/index.vue [hunk 1]
```diff
@@ -123,6 +132,26 @@ export default {
         this.$toast.error(err.message)
       }
     },
+    calculateItemHeight() {
```
Comment (roschaefer) [inline webapp/components/PostCard/index.vue]: @alina-beck do you still need this? The build server is crashing during frontend testing and it's trying to set some `style` attribute of `null`.

File: webapp/components/MasonryGrid/MasonryGridItem.vue [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+<script>
+export default {
+  data() {
+    return {
+      rowSpan: 10,
+    }
+  },
+  methods: {
+    calculateItemHeight() {
+      this.$parent.$emit('calculating-item-height')
```
Comment (mattwr18) [inline webapp/components/MasonryGrid/MasonryGridItem.vue]: I was curious what the difference between `this.$emit` and `this.$parent.$emit` was, and found this SO post.
As far as I understand it, this is the way to tell the parent component to emit an event to its parent? so like emit an event to a grandparent component.
Is that correct @alina-beck?

File: webapp/components/MasonryGrid/MasonryGridItem.vue [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      })
+    },
+  },
+  mounted() {
+    const image = this.$el.querySelector('img')
+    if (image) {
+      image.onload = () => this.calculateItemHeight()
+    } else {
+      // use timeout to make sure layout is set up before calculation
+      setTimeout(() => this.calculateItemHeight(), 0)
```
Comment (mattwr18) [inline webapp/components/MasonryGrid/MasonryGridItem.vue]: I like that you added a comment to explain this, but if the timeout is `0`, I'm a little confused what this does

File: webapp/pages/index.vue [hunk 4]
```diff
@@ -1,10 +1,10 @@
 <template>
   <div>
-    <ds-flex :width="{ base: '100%' }" gutter="base">
-      <ds-flex-item>
+    <masonry-grid>
+      <ds-grid-item v-show="hashtag" :row-span="2" column-span="fullWidth">
```
Comment (mattwr18) [inline webapp/pages/index.vue]: I was wondering where the `hashtag` was :smiley_cat:

File: webapp/components/MasonryGrid/MasonryGridItem.vue (hunk 2 above)
Comment (alina-beck) [reply webapp/components/MasonryGrid/MasonryGridItem.vue]: Yes, and the reason we have to do it here is that this component sits within a slot – I found the solution in [this comment](https://github.com/vuejs/vue/issues/4332#issuecomment-263444492) and it's working! But hard to test... 😕

File: webapp/components/MasonryGrid/MasonryGridItem.vue (hunk 3 above)
Comment (alina-beck) [reply webapp/components/MasonryGrid/MasonryGridItem.vue]: This has to do with the [JavaScript event loop](https://stackoverflow.com/questions/33955650/what-is-settimeout-doing-when-set-to-0-milliseconds/33955673). The timeout – even though it is set to `0` – sends the execution of this function to the back of the queue.

I am not sure which function _exactly_ we are waiting for because I don't know how Vue.js works its magic in the background but I know that without the timeout the layout is messed up. :smile:

**Comment by alina-beck** [comment]
@mattwr18 thanks for the review!

I noticed the hashtag issue – and I believe it has nothing to do with this PR. The first thing I did after Rob rebased my commits was test the hashtag feature (because I needed to toggle the `FilterMenu` and wanted to know what was going on) and already then realised that I could click on a hashtag but wouldn't see all the relevant posts...

**Comment by alina-beck** [comment]
@roschaefer Good catch! Thanks to you I found out that 

> CSS Grid Layout initially defined the `grid-row-gap` property. This prefixed property is being replaced by `row-gap`. (See [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/row-gap))

and am now using the new property (with the old one as a fallback). 😎

File: webapp/components/MasonryGrid/MasonryGridItem.vue (hunk 2 above)
Comment (roschaefer) [reply webapp/components/MasonryGrid/MasonryGridItem.vue]: You could emit again in the parent component
(child)-emit->(parent)-emit->(granddaddy)

File: webapp/components/MasonryGrid/MasonryGrid.vue [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<template>
+  <ds-grid
+    :min-column-width="300"
+    v-on:calculating-item-height="startCalculation"
```
Comment (roschaefer) [inline webapp/components/MasonryGrid/MasonryGrid.vue]:
```suggestion
    @calculating-item-height="startCalculation"
```
